File: sdks/js/src/providers/openai.ts
import { SpanKind, SpanStatusCode } from '@opentelemetry/api';
import { ATTR_LLM_MODEL, ATTR_LLM_PROVIDER } from '../attributes';
import type {
  LLMOperationType,
  LLMRequest,
  LLMResponse,
  LLMSpanAttributes,
} from '../types';
import { BaseProviderInstrumentation } from './base';

interface OpenAIClient {
  chat?: {
    completions?: {
      create: (params: OpenAIParams) => Promise<OpenAIResponse>;
    };
  };
  completions?: {
    create: (params: OpenAIParams) => Promise<OpenAIResponse>;
  };
  embeddings?: {
    create: (params: OpenAIParams) => Promise<OpenAIResponse>;
  };
  images?: {
    generate: (params: OpenAIParams) => Promise<OpenAIResponse>;
  };
  audio?: {
    transcriptions?: {
      create: (params: OpenAIParams) => Promise<OpenAIResponse>;
    };
  };
  moderations?: {
    create: (params: OpenAIParams) => Promise<OpenAIResponse>;
  };
}

interface OpenAIParams {
  model?: string;
  messages?: Array<{ role: string; content: string }>;
  prompt?: string | string[];
  input?: string | string[];
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  stream?: boolean;
  tools?: Array<{ name: string; description?: string; parameters?: unknown }>;
}

interface OpenAIResponse {
  id?: string;
  model?: string;
  choices?: Array<{
    text?: string;
    message?: { role: string; content: string };
    index: number;
    finish_reason?: string;
  }>;
  data?: Array<{ embedding: number[] }>;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
}

/**
 * OpenAI provider instrumentation
 */
export class OpenAIInstrumentation extends BaseProviderInstrumentation {
  // biome-ignore lint/complexity/noBannedTypes: This is a valid use of any
  private originalMethods: Map<string, Function> = new Map();

  constructor() {
    super('openai', '1.0.0');
  }

  canInstrument(module: unknown): boolean {
    // Check if this is an OpenAI client
    return (
      module !== null &&
      typeof module === 'object' &&
      ('chat' in module || 'completions' in module || 'embeddings' in module)
    );
  }

  instrument<T = unknown>(module: T): T {
    const client = module as OpenAIClient;

    this.debug('Instrumenting OpenAI client:', {
      hasChat: !!client.chat,
      hasCompletions: !!client.completions,
      hasEmbeddings: !!client.embeddings,
    });

    // Instrument chat completions
    if (client.chat?.completions?.create) {
      this.debug('Instrumenting chat.completions.create');
      const originalCreate = client.chat.completions.create.bind(
        client.chat.completions,
      );
      this.originalMethods.set('chat.completions.create', originalCreate);
      client.chat.completions.create = this.wrapMethod(
        originalCreate,
        'chat',
        'chat.completions.create',
      );
    }

    // Instrument completions
    if (client.completions?.create) {
      const originalCreate = client.completions.create.bind(client.completions);
      this.originalMethods.set('completions.create', originalCreate);
      client.completions.create = this.wrapMethod(
        originalCreate,
        'completion',
        'completions.create',
      );
    }

    // Instrument embeddings
    if (client.embeddings?.create) {
      const originalCreate = client.embeddings.create.bind(client.embeddings);
      this.originalMethods.set('embeddings.create', originalCreate);
      client.embeddings.create = this.wrapMethod(
        originalCreate,
        'embedding',
        'embeddings.create',
      );
    }

    // Instrument images
    if (client.images?.generate) {
      const originalGenerate = client.images.generate.bind(client.images);
      this.originalMethods.set('images.generate', originalGenerate);
      client.images.generate = this.wrapMethod(
        originalGenerate,
        'image_generation',
        'images.generate',
      );
    }

    // Instrument audio
    if (client.audio?.transcriptions?.create) {
      const originalCreate = client.audio.transcriptions.create.bind(
        client.audio.transcriptions,
      );
      this.originalMethods.set('audio.transcriptions.create', originalCreate);
      client.audio.transcriptions.create = this.wrapMethod(
        originalCreate,
        'audio_transcription',
        'audio.transcriptions.create',
      );
    }

    // Instrument moderations
    if (client.moderations?.create) {
      const originalCreate = client.moderations.create.bind(client.moderations);
      this.originalMethods.set('moderations.create', originalCreate);
      client.moderations.create = this.wrapMethod(
        originalCreate,
        'moderation',
        'moderations.create',
      );
    }

    return module;
  }

  enable(): void {
    // Enable instrumentation - this is called when the instrumentation is enabled
    this.debug('OpenAI instrumentation enabled');
  }

  disable(): void {
    // Restore original methods
    this.originalMethods.forEach((_originalMethod, methodName) => {
      // This would need to be implemented based on how we track the instrumented modules
      this.debug(`Restored original method: ${methodName}`);
    });
    this.originalMethods.clear();
    this.debug('OpenAI instrumentation disabled');
  }

  private wrapMethod(
    originalMethod: (params: OpenAIParams) => Promise<OpenAIResponse>,
    operationType: LLMOperationType,
    methodName: string,
  ) {
    const instrumentation = this;

    return async function (
      this: unknown,
      params: OpenAIParams,
    ): Promise<OpenAIResponse> {
      instrumentation.debug(`Executing wrapped method: ${methodName}`, {
        messagesCount: params.messages?.length,
        model: params.model,
        operationType,
      });

      const tracer = instrumentation.getTracer();
      const span = tracer.startSpan(`openai.${methodName}`, {
        kind: SpanKind.CLIENT,
      });

      const startTime = Date.now();

      try {
        // Extract request data
        const request = instrumentation.extractRequest(params, operationType);

        // Set initial attributes
        const attributes: LLMSpanAttributes = {
          [ATTR_LLM_PROVIDER]: 'openai',
          [ATTR_LLM_MODEL]: request.model,
          'llm.operation.type': operationType,
        };

        if (request.parameters?.temperature !== undefined) {
          attributes['llm.temperature'] = request.parameters
            .temperature as number;
        }
        if (request.parameters?.max_tokens !== undefined) {
          attributes['llm.max_tokens'] = request.parameters
            .max_tokens as number;
        }
        if (request.parameters?.stream !== undefined) {
          attributes['llm.stream'] = request.parameters.stream as boolean;
        }

        // Add messages for chat operations
        if (request.messages) {
          attributes['llm.messages'] = JSON.stringify(request.messages);
        }

        // Add prompt for completion operations
        if (request.prompt) {
          attributes['llm.prompt'] =
            typeof request.prompt === 'string'
              ? request.prompt
              : JSON.stringify(request.prompt);
        }

        // Add input for embedding operations
        if (request.input) {
          attributes['llm.input'] =
            typeof request.input === 'string'
              ? request.input
              : JSON.stringify(request.input);
        }

        // Add tools if present
        if (request.tools) {
          attributes['llm.tools'] = JSON.stringify(request.tools);
        }

        span.setAttributes(attributes);

        // Execute original method
        const result = await originalMethod.call(this, params);

        // Extract response data
        const response = instrumentation.extractResponse(result, operationType);

        // Update attributes with response data
        if (response.usage) {
          if (response.usage.promptTokens !== undefined) {
            span.setAttribute('llm.prompt.tokens', response.usage.promptTokens);
          }
          if (response.usage.completionTokens !== undefined) {
            span.setAttribute(
              'llm.completion.tokens',
              response.usage.completionTokens,
            );
          }
          if (response.usage.totalTokens !== undefined) {
            span.setAttribute('llm.total.tokens', response.usage.totalTokens);
          }
        }

        if (response.requestId) {
          span.setAttribute('llm.request.id', response.requestId);
        }

        // Add response choices for chat/completion operations
        if (response.choices) {
          span.setAttribute('llm.choices', JSON.stringify(response.choices));
        }

        // Add embeddings for embedding operations
        if (response.embeddings) {
          span.setAttribute(
            'llm.embeddings',
            JSON.stringify(response.embeddings),
          );
        }

        const duration = Date.now() - startTime;
        span.setAttribute('llm.duration_ms', duration);

        // Calculate costs if available
        const cost = instrumentation.calculateCost(
          request.model,
          response.usage,
        );
        if (cost) {
          if (cost.prompt !== undefined) {
            span.setAttribute('llm.cost.prompt', cost.prompt);
          }
          if (cost.completion !== undefined) {
            span.setAttribute('llm.cost.completion', cost.completion);
          }
          span.setAttribute('llm.cost.total', cost.total);
        }

        span.setStatus({ code: SpanStatusCode.OK });
        span.end();

        return result;
      } catch (error) {
        // Record error
        span.recordException(error as Error);
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: (error as Error).message,
        });

        if (error instanceof Error) {
          span.setAttribute('llm.error', error.message);
          span.setAttribute('llm.error.type', error.name);
        }

        span.end();
        throw error;
      }
    };
  }

  private extractRequest(
    params: OpenAIParams,
    operationType: LLMOperationType,
  ): LLMRequest {
    const request: LLMRequest = {
      model: params.model || 'unknown',
      operation: operationType,
      parameters: {},
      provider: 'openai',
    };

    // Extract common parameters
    if (params.temperature !== undefined && request.parameters) {
      request.parameters.temperature = params.temperature;
    }
    if (params.max_tokens !== undefined && request.parameters) {
      request.parameters.max_tokens = params.max_tokens;
    }
    if (params.top_p !== undefined && request.parameters) {
      request.parameters.top_p = params.top_p;
    }
    if (params.stream !== undefined && request.parameters) {
      request.parameters.stream = params.stream;
    }

    // Extract operation-specific data
    switch (operationType) {
      case 'chat':
        request.messages = params.messages;
        if (params.tools) {
          request.tools = params.tools;
        }
        break;
      case 'completion':
        request.prompt = params.prompt;
        break;
      case 'embedding':
        request.input = params.input;
        break;
    }

    return request;
  }

  private extractResponse(
    response: OpenAIResponse,
    operationType: LLMOperationType,
  ): LLMResponse {
    const result: LLMResponse = {
      model: response.model || 'unknown',
    };

    // Extract common fields
    if (response.id) {
      result.requestId = response.id;
    }

    if (response.usage) {
      result.usage = {
        completionTokens: response.usage.completion_tokens,
        promptTokens: response.usage.prompt_tokens,
        totalTokens: response.usage.total_tokens,
      };
    }

    // Extract operation-specific data
    switch (operationType) {
      case 'chat':
      case 'completion':
        if (response.choices) {
          result.choices = response.choices;
        }
        break;
      case 'embedding':
        if (response.data) {
          result.embeddings = response.data.map((item) => item.embedding);
        }
        break;
    }

    return result;
  }

  private calculateCost(
    model: string,
    usage?: { promptTokens?: number; completionTokens?: number },
  ) {
    if (!usage || !usage.promptTokens || !usage.completionTokens) {
      return null;
    }

    // Cost per 1K tokens (example rates, should be configurable)
    const costs: Record<string, { prompt: number; completion: number }> = {
      'gpt-3.5-turbo': { completion: 0.0015, prompt: 0.0005 },
      'gpt-4': { completion: 0.06, prompt: 0.03 },
      'gpt-4-turbo': { completion: 0.03, prompt: 0.01 },
      'text-embedding-ada-002': { completion: 0, prompt: 0.0001 },
    };

    const modelCost = costs[model];
    if (!modelCost) {
      return null;
    }

    const promptCost = (usage.promptTokens / 1000) * modelCost.prompt;
    const completionCost =
      (usage.completionTokens / 1000) * modelCost.completion;

    return {
      completion: completionCost,
      prompt: promptCost,
      total: promptCost + completionCost,
    };
  }
}

File: sdks/python/src/untrace/instrumentation/openai.py
"""OpenAI instrumentation implementation."""

import functools
import logging
from typing import Any, Callable, Dict

from .base import BaseProviderInstrumentation

logger = logging.getLogger(__name__)


class OpenAIInstrumentation(BaseProviderInstrumentation):
    """OpenAI SDK instrumentation."""

    def __init__(self, tracer, metrics, context):
        """Initialize OpenAI instrumentation."""
        super().__init__(tracer, metrics, context, "openai")
        self._original_constructors = {}

    def instrument(self) -> None:
        """Instrument OpenAI client methods."""
        if self._instrumented:
            return

        try:
            # Import OpenAI here to avoid import errors if not installed
            from openai import OpenAI, AsyncOpenAI

            # Store original constructors
            self._original_constructors['OpenAI'] = OpenAI.__init__
            self._original_constructors['AsyncOpenAI'] = AsyncOpenAI.__init__

            # Patch constructors to instrument instances
            OpenAI.__init__ = self._wrap_openai_constructor(OpenAI.__init__)
            AsyncOpenAI.__init__ = self._wrap_async_openai_constructor(AsyncOpenAI.__init__)

            self._mark_instrumented()

        except ImportError:
            logger.warning("[Untrace] OpenAI not installed, skipping instrumentation")

    def _wrap_openai_constructor(self, original_constructor):
        """Wrap OpenAI constructor to instrument the instance."""
        instrumentation = self

        @functools.wraps(original_constructor)
        def wrapper(self, *args, **kwargs):
            # Call original constructor
            original_constructor(self, *args, **kwargs)

            # Instrument the instance methods
            instrumentation._instrument_client_instance(self)

        return wrapper

    def _wrap_async_openai_constructor(self, original_constructor):
        """Wrap AsyncOpenAI constructor to instrument the instance."""
        instrumentation = self

        @functools.wraps(original_constructor)
        def wrapper(self, *args, **kwargs):
            # Call original constructor
            original_constructor(self, *args, **kwargs)

            # Instrument the instance methods
            instrumentation._instrument_client_instance(self)

        return wrapper

    def _instrument_client_instance(self, client_instance):
        """Instrument a client instance."""
        # Instrument chat completions
        if hasattr(client_instance, 'chat') and hasattr(client_instance.chat, 'completions'):
            self._patch_method(
                client_instance.chat.completions,
                'create',
                self._wrap_chat_completion_sync
            )

        # Instrument completions
        if hasattr(client_instance, 'completions'):
            self._patch_method(
                client_instance.completions,
                'create',
                self._wrap_completion_sync
            )

        # Instrument embeddings
        if hasattr(client_instance, 'embeddings'):
            self._patch_method(
                client_instance.embeddings,
                'create',
                self._wrap_embedding_sync
            )

    def _wrap_chat_completion_sync(self, original_method):
        """Wrap synchronous chat completion method."""
        @functools.wraps(original_method)
        def wrapper(*args, **kwargs):
            # Extract model and messages from kwargs
            model = self._extract_model_from_kwargs(kwargs)
            messages = self._extract_messages_from_kwargs(kwargs)

            # Create LLM attributes
            attributes = self._create_llm_attributes(
                model=model,
                operation_type="chat",
                temperature=kwargs.get('temperature'),
                max_tokens=kwargs.get('max_tokens'),
                stream=kwargs.get('stream', False)
            )

            # Create span name
            span_name = f"openai.chat.completions.create"

            # Wrap the call with tracing
            def traced_call(span):
                # Set request attributes
                self._set_request_attributes(span, model, **kwargs)
                span.set_attribute("llm.request.messages", str(messages))

                span.add_event("llm.request.start")

                # Make the original API call
                response = original_method(*args, **kwargs)

                span.add_event("llm.response.received")

                # Extract response data
                if hasattr(response, 'choices') and response.choices:
                    choice = response.choices[0]
                    if hasattr(choice, 'message'):
                        span.set_attribute("llm.response.content", choice.message.content)
                    elif hasattr(choice, 'text'):
                        span.set_attribute("llm.response.content", choice.text)

                # Set response attributes
                self._set_response_attributes(span, response)

                # Handle usage information
                self._handle_usage_metrics(span, response, model)

                return response

            return self._create_llm_span_sync(span_name, attributes, traced_call)

        return wrapper

    def _wrap_completion_sync(self, original_method):
        """Wrap synchronous completion method."""
        @functools.wraps(original_method)
        def wrapper(*args, **kwargs):
            # Extract model and prompt from kwargs
            model = self._extract_model_from_kwargs(kwargs)
            prompt = self._extract_prompt_from_kwargs(kwargs)

            # Create LLM attributes
            attributes = self._create_llm_attributes(
                model=model,
                operation_type="completion",
                temperature=kwargs.get('temperature'),
                max_tokens=kwargs.get('max_tokens')
            )

            # Create span name
            span_name = f"openai.completions.create"

            # Wrap the call with tracing
            def traced_call(span):
                # Set request attributes
                self._set_request_attributes(span, model, **kwargs)
                span.set_attribute("llm.request.prompt", prompt)

                span.add_event("llm.request.start")

                # Make the original API call
                response = original_method(*args, **kwargs)

                span.add_event("llm.response.received")

                # Extract response data
                if hasattr(response, 'choices') and response.choices:
                    choice = response.choices[0]
                    if hasattr(choice, 'text'):
                        span.set_attribute("llm.response.text", choice.text)

                # Set response attributes
                self._set_response_attributes(span, response)

                # Handle usage information
                self._handle_usage_metrics(span, response, model)

                return response

            return self._create_llm_span_sync(span_name, attributes, traced_call)

        return wrapper

    def _wrap_embedding_sync(self, original_method):
        """Wrap synchronous embedding method."""
        @functools.wraps(original_method)
        def wrapper(*args, **kwargs):
            # Extract model and input from kwargs
            model = self._extract_model_from_kwargs(kwargs)
            input_text = self._extract_input_from_kwargs(kwargs)

            # Create LLM attributes
            attributes = self._create_llm_attributes(
                model=model,
                operation_type="embedding"
            )

            # Create span name
            span_name = f"openai.embeddings.create"

            # Wrap the call with tracing
            def traced_call(span):
                # Set request attributes
                self._set_request_attributes(span, model, **kwargs)
                span.set_attribute("llm.request.input", str(input_text))

                span.add_event("llm.request.start")

                # Make the original API call
                response = original_method(*args, **kwargs)

                span.add_event("llm.response.received")

                # Extract response data
                if hasattr(response, 'data') and response.data:
                    embedding = response.data[0]
                    span.set_attribute("llm.response.embeddings_count", len(response.data))
                    span.set_attribute("llm.response.embedding_dimension", len(embedding.embedding))

                # Set response attributes
                self._set_response_attributes(span, response)

                # Handle usage information
                self._handle_usage_metrics(span, response, model)

                return response

            return self._create_llm_span_sync(span_name, attributes, traced_call)

        return wrapper

    def uninstrument(self) -> None:
        """Remove instrumentation."""
        if not self._instrumented:
            return

        # Restore original constructors
        try:
            from openai import OpenAI, AsyncOpenAI
            OpenAI.__init__ = self._original_constructors['OpenAI']
            AsyncOpenAI.__init__ = self._original_constructors['AsyncOpenAI']
        except ImportError:
            pass

        self._mark_uninstrumented()
